### PR TITLE
New RPGLoot Affixes + Component Additions

### PR DESCRIPTION
**Pull request by Whoneedspacee**
## About The Pull Request

Adds fired_from to projectiles to store the thing the projectile was fired from, firer = human that fired, fired_from = the gun that shot the projectile. Knockback now handles projectiles knocking back and also anchored structures being thrown.

Adds shrapnel component, a component that shoots projectiles in a radius around a projectile when it hits something. Adds igniter component, sets mobs the parent hits on fire. Adds summoning component, summons mobs when the parent hits something. Adds lifesteal component that attempts to heal the attacker for an amount when they hit something living.

Also adds some ishelpers and signals

## Why It's Good For The Game

RPGLoot is pretty lacking in actual components that do anything and for the most part are completely cosmetic. This doesn't totally address the imbalance of cosmetic affixes that do absolutely nothing, but at least adds some more variety other than 2 things that can occur.

## Changelog
:cl:
add: New RPGLoot modifiers: Vampirism which heals you when you attack, Pyromantic which sets things you hit on fire. Shrapnel which causes projectiles fired from a gun to fire projectiles in a radius when they hit something. Finally, Summoning which summons mobs that sometimes aid you in combat.
/:cl:

### CODE REVIEW

**Comment by Militaires** [comment]
good pr

**Comment by Razharas** [comment]
yeah but what about adding the stats and resistances to weapons and armour?
how am i supposed to fight high tier boss content a if only options are for white/blue clearspeed?

**Comment by 4dplanner** [comment]
If there are going to be more of these fantasy components, maybe they should get their own folder?

**Comment by Whoneedspacee** [comment]
they aren’t specifically for fantasy they are components.

the fantasy component already has its own folder which applies these components in the specific way they should be for rpgloot.

**Comment by 4dplanner** [comment]
lol I actually just misread the path on the fantasy components, ignore me

File: code/datums/components/fantasy/suffixes.dm [hunk 1]
```diff
@@ -78,14 +77,76 @@
...
+/datum/fantasy_affix/shrapnel
+	placement = AFFIX_SUFFIX
+	alignment = AFFIX_GOOD
+
+/datum/fantasy_affix/shrapnel/apply(datum/component/fantasy/comp, newName)
+	. = ..()
+	var/static/list/possible_projectiletypes
+	if(!possible_projectiletypes)
+		// The base list of allowed types
+		possible_projectiletypes = typecacheof(list(
```
Comment (ninjanomnom) [inline code/datums/components/fantasy/suffixes.dm]: This should probably be hand picked and not just every kind, they have a *lot* of variety and many will break things.

This also lets you weight them if you want

File: code/datums/components/knockback.dm [hunk 2]
```diff
@@ -1,22 +1,45 @@
 /datum/component/knockback
 	var/throw_distance
+	var/reverse
```
Comment (ninjanomnom) [inline code/datums/components/knockback.dm]: Just have a negative throw distance?

**Review by ninjanomnom — CHANGES_REQUESTED** [review]

File: code/datums/components/knockback.dm (hunk 2 above)
Comment (Whoneedspacee) [reply code/datums/components/knockback.dm]: oh duh haha i got stuck up in the fact that getting negative target turfs wouldnt work

**Pull request merged**